File: AIP_Backend/Services/SiteService.cs
#nullable enable

using AIPBackend.Data;
using AIPBackend.Models;
using AIPBackend.Models.DTOs;
using Microsoft.EntityFrameworkCore;

namespace AIPBackend.Services
{
    public class SiteService : ISiteService
    {
        private readonly ApplicationDbContext _context;
        private readonly IUserContextService _userContext;

        public SiteService(ApplicationDbContext context, IUserContextService userContext)
        {
            _context = context;
            _userContext = userContext;
        }

        public async Task<List<SiteDto>> GetSitesAsync(int page = 1, int pageSize = 10, string? search = null, int? customerId = null, int? regionId = null)
        {
            var context = _userContext.GetCurrentContext();
            var customerFilter = _userContext.ResolveCustomerFilter(customerId);
            var query = _context.Sites.AsQueryable().ApplyTenantScope(customerFilter, context);

            // Apply filters
            if (!string.IsNullOrEmpty(search))
            {
                query = query.Where(s => s.LocationName.Contains(search) || 
                                        s.BuildingName!.Contains(search) || 
                                        s.Town!.Contains(search) || 
                                        s.County!.Contains(search));
            }

            if (regionId.HasValue)
            {
                query = query.Where(s => s.fkRegionID == regionId.Value);
            }

            // Only return non-deleted sites
            query = query.Where(s => !s.RecordIsDeletedYN);

            var sites = await query
                .OrderBy(s => s.LocationName) // Add default ordering to fix EF warning
                .Skip((page - 1) * pageSize)
                .Take(pageSize)
                .Select(s => new SiteDto
                {
                    SiteID = s.SiteID,
                    fkCustomerID = s.fkCustomerID,
                    fkRegionID = s.fkRegionID,
                    CoreSiteYN = s.CoreSiteYN,
                    LocationName = s.LocationName,
                    SINNumber = s.SINNumber,
                    LocationType = s.LocationType,
                    BuildingName = s.BuildingName,
                    NumberandStreet = s.NumberandStreet,
                    VillageOrSuburb = s.VillageOrSuburb,
                    Town = s.Town,
                    County = s.County,
                    Postcode = s.Postcode,
                    TelephoneNumber = s.TelephoneNumber,
                    ContractStartDate = s.ContractStartDate,
                    ContractEndDate = s.ContractEndDate,
                    Details = s.Details,
                    SiteSurveyComplete = s.SiteSurveyComplete,
                    AssignmentInstructionsIssued = s.AssignmentInstructionsIssued,
                    RiskAssessmentIssued = s.RiskAssessmentIssued,
                    RecordIsDeletedYN = s.RecordIsDeletedYN,
                    DateCreated = s.DateCreated,
                    CreatedBy = s.CreatedBy,
                    DateModified = s.DateModified,
                    ModifiedBy = s.ModifiedBy
                })
                .ToListAsync();

            return sites;
        }

        public async Task<SiteDto?> GetSiteByIdAsync(int id)
        {
            var site = await _context.Sites
                .FirstOrDefaultAsync(s => s.SiteID == id && !s.RecordIsDeletedYN);

            if (site == null)
                return null;

            _userContext.EnsureCanAccessCustomer(site.fkCustomerID);
            var userContext = _userContext.GetCurrentContext();
            if (userContext.AccessibleSiteIds.Count > 0 &&
                !userContext.AccessibleSiteIds.Contains(site.SiteID.ToString()))
            {
                throw new AIPBackend.Exceptions.ForbiddenAccessException("You do not have permission to access this site.");
            }

            return new SiteDto
            {
                SiteID = site.SiteID,
                fkCustomerID = site.fkCustomerID,
                fkRegionID = site.fkRegionID,
                CoreSiteYN = site.CoreSiteYN,
                LocationName = site.LocationName,
                SINNumber = site.SINNumber,
                LocationType = site.LocationType,
                BuildingName = site.BuildingName,
                NumberandStreet = site.NumberandStreet,
                VillageOrSuburb = site.VillageOrSuburb,
                Town = site.Town,
                County = site.County,
                Postcode = site.Postcode,
                TelephoneNumber = site.TelephoneNumber,
                ContractStartDate = site.ContractStartDate,
                ContractEndDate = site.ContractEndDate,
                Details = site.Details,
                SiteSurveyComplete = site.SiteSurveyComplete,
                AssignmentInstructionsIssued = site.AssignmentInstructionsIssued,
                RiskAssessmentIssued = site.RiskAssessmentIssued,
                RecordIsDeletedYN = site.RecordIsDeletedYN,
                DateCreated = site.DateCreated,
                CreatedBy = site.CreatedBy,
                DateModified = site.DateModified,
                ModifiedBy = site.ModifiedBy
            };
        }

        public async Task<List<SiteDto>> GetSitesByCustomerAsync(int customerId)
        {
            var context = _userContext.GetCurrentContext();
            var customerFilter = _userContext.ResolveCustomerFilter(customerId);
            var sites = await _context.Sites
                .Where(s => !s.RecordIsDeletedYN)
                .ApplyTenantScope(customerFilter, context)
                .Select(s => new SiteDto
                {
                    SiteID = s.SiteID,
                    fkCustomerID = s.fkCustomerID,
                    fkRegionID = s.fkRegionID,
                    CoreSiteYN = s.CoreSiteYN,
                    LocationName = s.LocationName,
                    SINNumber = s.SINNumber,
                    LocationType = s.LocationType,
                    BuildingName = s.BuildingName,
                    NumberandStreet = s.NumberandStreet,
                    VillageOrSuburb = s.VillageOrSuburb,
                    Town = s.Town,
                    County = s.County,
                    Postcode = s.Postcode,
                    TelephoneNumber = s.TelephoneNumber,
                    ContractStartDate = s.ContractStartDate,
                    ContractEndDate = s.ContractEndDate,
                    Details = s.Details,
                    SiteSurveyComplete = s.SiteSurveyComplete,
                    AssignmentInstructionsIssued = s.AssignmentInstructionsIssued,
                    RiskAssessmentIssued = s.RiskAssessmentIssued,
                    RecordIsDeletedYN = s.RecordIsDeletedYN,
                    DateCreated = s.DateCreated,
                    CreatedBy = s.CreatedBy,
                    DateModified = s.DateModified,
                    ModifiedBy = s.ModifiedBy
                })
                .ToListAsync();

            return sites;
        }

        public async Task<List<SiteDto>> GetSitesByRegionAsync(int regionId)
        {
            var region = await _context.Regions
                .AsNoTracking()
                .FirstOrDefaultAsync(r => r.RegionID == regionId && !r.RecordIsDeletedYN);
            if (region == null)
            {
                return new List<SiteDto>();
            }

            _userContext.EnsureCanAccessCustomer(region.fkCustomerID);
            var context = _userContext.GetCurrentContext();
            var customerFilter = _userContext.ResolveCustomerFilter(region.fkCustomerID);
            var sites = await _context.Sites
                .Where(s => s.fkRegionID == regionId && !s.RecordIsDeletedYN)
                .ApplyTenantScope(customerFilter, context)
                .Select(s => new SiteDto
                {
                    SiteID = s.SiteID,
                    fkCustomerID = s.fkCustomerID,
                    fkRegionID = s.fkRegionID,
                    CoreSiteYN = s.CoreSiteYN,
                    LocationName = s.LocationName,
                    SINNumber = s.SINNumber,
                    LocationType = s.LocationType,
                    BuildingName = s.BuildingName,
                    NumberandStreet = s.NumberandStreet,
                    VillageOrSuburb = s.VillageOrSuburb,
                    Town = s.Town,
                    County = s.County,
                    Postcode = s.Postcode,
                    TelephoneNumber = s.TelephoneNumber,
                    ContractStartDate = s.ContractStartDate,
                    ContractEndDate = s.ContractEndDate,
                    Details = s.Details,
                    SiteSurveyComplete = s.SiteSurveyComplete,
                    AssignmentInstructionsIssued = s.AssignmentInstructionsIssued,
                    RiskAssessmentIssued = s.RiskAssessmentIssued,
                    RecordIsDeletedYN = s.RecordIsDeletedYN,
                    DateCreated = s.DateCreated,
                    CreatedBy = s.CreatedBy,
                    DateModified = s.DateModified,
                    ModifiedBy = s.ModifiedBy
                })
                .ToListAsync();

            return sites;
        }

        public async Task<SiteDto> CreateSiteAsync(SiteCreateRequestDto createSiteDto, string createdBy)
        {
            _userContext.EnsureCanAccessCustomer(createSiteDto.fkCustomerID);
            var site = new Site
            {
                fkCustomerID = createSiteDto.fkCustomerID,
                fkRegionID = createSiteDto.fkRegionID,
                CoreSiteYN = createSiteDto.CoreSiteYN,
                LocationName = createSiteDto.LocationName,
                SINNumber = createSiteDto.SINNumber,
                LocationType = createSiteDto.LocationType,
                BuildingName = createSiteDto.BuildingName,
                NumberandStreet = createSiteDto.NumberandStreet,
                VillageOrSuburb = createSiteDto.VillageOrSuburb,
                Town = createSiteDto.Town,
                County = createSiteDto.County,
                Postcode = createSiteDto.Postcode,
                TelephoneNumber = createSiteDto.TelephoneNumber,
                ContractStartDate = createSiteDto.ContractStartDate,
                ContractEndDate = createSiteDto.ContractEndDate,
                Details = createSiteDto.Details,
                SiteSurveyComplete = createSiteDto.SiteSurveyComplete,
                AssignmentInstructionsIssued = createSiteDto.AssignmentInstructionsIssued,
                RiskAssessmentIssued = createSiteDto.RiskAssessmentIssued,
                RecordIsDeletedYN = false,
                DateCreated = DateTime.UtcNow,
                CreatedBy = createdBy,
                DateModified = null,
                ModifiedBy = null
            };

            _context.Sites.Add(site);
            await _context.SaveChangesAsync();

            return new SiteDto
            {
                SiteID = site.SiteID,
                fkCustomerID = site.fkCustomerID,
                fkRegionID = site.fkRegionID,
                CoreSiteYN = site.CoreSiteYN,
                LocationName = site.LocationName,
                SINNumber = site.SINNumber,
                LocationType = site.LocationType,
                BuildingName = site.BuildingName,
                NumberandStreet = site.NumberandStreet,
                VillageOrSuburb = site.VillageOrSuburb,
                Town = site.Town,
                County = site.County,
                Postcode = site.Postcode,
                TelephoneNumber = site.TelephoneNumber,
                ContractStartDate = site.ContractStartDate,
                ContractEndDate = site.ContractEndDate,
                Details = site.Details,
                SiteSurveyComplete = site.SiteSurveyComplete,
                AssignmentInstructionsIssued = site.AssignmentInstructionsIssued,
                RiskAssessmentIssued = site.RiskAssessmentIssued,
                RecordIsDeletedYN = site.RecordIsDeletedYN,
                DateCreated = site.DateCreated,
                CreatedBy = site.CreatedBy,
                DateModified = site.DateModified,
                ModifiedBy = site.ModifiedBy
            };
        }

        public async Task<SiteDto?> UpdateSiteAsync(int id, SiteUpdateRequestDto updateSiteDto, string modifiedBy)
        {
            var site = await _context.Sites
                .FirstOrDefaultAsync(s => s.SiteID == id && !s.RecordIsDeletedYN);

            if (site == null)
                return null;

            _userContext.EnsureCanAccessCustomer(site.fkCustomerID);
            site.CoreSiteYN = updateSiteDto.CoreSiteYN;
            site.LocationName = updateSiteDto.LocationName ?? site.LocationName;
            site.SINNumber = updateSiteDto.SINNumber;
            site.LocationType = updateSiteDto.LocationType;
            site.BuildingName = updateSiteDto.BuildingName;
            site.NumberandStreet = updateSiteDto.NumberandStreet;
            site.VillageOrSuburb = updateSiteDto.VillageOrSuburb;
            site.Town = updateSiteDto.Town;
            site.County = updateSiteDto.County;
            site.Postcode = updateSiteDto.Postcode;
            site.TelephoneNumber = updateSiteDto.TelephoneNumber;
            site.ContractStartDate = updateSiteDto.ContractStartDate;
            site.ContractEndDate = updateSiteDto.ContractEndDate;
            site.Details = updateSiteDto.Details;
            site.SiteSurveyComplete = updateSiteDto.SiteSurveyComplete;
            site.AssignmentInstructionsIssued = updateSiteDto.AssignmentInstructionsIssued;
            site.RiskAssessmentIssued = updateSiteDto.RiskAssessmentIssued;
            site.DateModified = DateTime.UtcNow;
            site.ModifiedBy = modifiedBy ?? "system";

            await _context.SaveChangesAsync();

            return new SiteDto
            {
                SiteID = site.SiteID,
                fkCustomerID = site.fkCustomerID,
                fkRegionID = site.fkRegionID,
                CoreSiteYN = site.CoreSiteYN,
                LocationName = site.LocationName,
                SINNumber = site.SINNumber,
                LocationType = site.LocationType,
                BuildingName = site.BuildingName,
                NumberandStreet = site.NumberandStreet,
                VillageOrSuburb = site.VillageOrSuburb,
                Town = site.Town,
                County = site.County,
                Postcode = site.Postcode,
                TelephoneNumber = site.TelephoneNumber,
                ContractStartDate = site.ContractStartDate,
                ContractEndDate = site.ContractEndDate,
                Details = site.Details,
                SiteSurveyComplete = site.SiteSurveyComplete,
                AssignmentInstructionsIssued = site.AssignmentInstructionsIssued,
                RiskAssessmentIssued = site.RiskAssessmentIssued,
                RecordIsDeletedYN = site.RecordIsDeletedYN,
                DateCreated = site.DateCreated,
                CreatedBy = site.CreatedBy,
                DateModified = site.DateModified,
                ModifiedBy = site.ModifiedBy
            };
        }

        public async Task<bool> DeleteSiteAsync(int id)
        {
            var site = await _context.Sites
                .FirstOrDefaultAsync(s => s.SiteID == id && !s.RecordIsDeletedYN);

            if (site == null)
                return false;

            _userContext.EnsureCanAccessCustomer(site.fkCustomerID);
            // Soft delete
            site.RecordIsDeletedYN = true;
            site.DateModified = DateTime.UtcNow;
            site.ModifiedBy = "system"; // TODO: Get from current user

            await _context.SaveChangesAsync();
            return true;
        }
    }
}

File: AIP_UI/src/services/siteService.ts
import { api, SITE_ENDPOINTS, handleApiError } from '@/config/api'
import type { Site } from '@/types/customer'

export interface CreateSiteData {
  fkCustomerID: number
  fkRegionID: number
  coreSiteYN?: boolean
  locationName: string
  locationType?: string
  sinNumber?: string
  buildingName?: string
  numberandStreet?: string
  villageOrSuburb?: string
  town?: string
  county?: string
  postcode?: string
  telephoneNumber?: string
  contractStartDate?: string
  contractEndDate?: string
  details?: string
  siteSurveyComplete?: string
  assignmentInstructionsIssued?: string
  riskAssessmentIssued?: string
  recordIsDeletedYN?: boolean
  createdBy: string
}

export interface UpdateSiteData {
  coreSiteYN?: boolean
  locationName?: string
  locationType?: string
  sinNumber?: string
  buildingName?: string
  numberandStreet?: string
  villageOrSuburb?: string
  town?: string
  county?: string
  postcode?: string
  telephoneNumber?: string
  contractStartDate?: string
  contractEndDate?: string
  details?: string
  siteSurveyComplete?: string
  assignmentInstructionsIssued?: string
  riskAssessmentIssued?: string
  recordIsDeletedYN?: boolean
  modifiedBy: string
}

export interface ApiResponse<T> {
  success: boolean
  message: string
  data: T
  errors?: string[]
}

export const siteService = {
  // Get all sites
  async getSites(customerId?: number, regionId?: number): Promise<{ success: boolean; data: Site[] }> {
    try {
      console.log('🔄 [SiteService] Fetching sites from backend')
      
      const params = new URLSearchParams()
      if (customerId) {
        params.append('customerId', customerId.toString())
      }
      if (regionId) {
        params.append('regionId', regionId.toString())
      }
      
      const response = await api.get<ApiResponse<Site[]>>(`${SITE_ENDPOINTS.LIST}?${params}`)
      
      if (response.data.success) {
        console.log('✅ [SiteService] Successfully fetched sites:', response.data.data.length)
        return {
          success: true,
          data: response.data.data
        }
      } else {
        console.error('❌ [SiteService] Failed to fetch sites:', response.data.message)
        return {
          success: false,
          data: []
        }
      }
    } catch (error) {
      console.error('❌ [SiteService] Error fetching sites:', error)
      return {
        success: false,
        data: []
      }
    }
  },

  // Get sites by customer
  async getSitesByCustomer(customerId: number): Promise<{ success: boolean; data: Site[] }> {
    try {
      console.log('🔄 [SiteService] Fetching sites for customer:', customerId)
      
      const response = await api.get<ApiResponse<any[]>>(SITE_ENDPOINTS.BY_CUSTOMER(customerId.toString()))
      
      if (response.data.success) {
        // Map backend response to frontend Site type
        // Backend returns SiteID and LocationName (PascalCase), frontend expects siteID and locationName (camelCase)
        const mappedSites: Site[] = response.data.data.map((site: any) => ({
          siteID: site.SiteID || site.siteID || site.SiteId || site.siteId,
          fkCustomerID: site.fkCustomerID || site.fkCustomerId,
          fkRegionID: site.fkRegionID || site.fkRegionId,
          coreSiteYN: site.CoreSiteYN ?? site.coreSiteYN ?? false,
          locationName: site.LocationName || site.locationName || '',
          locationType: site.LocationType || site.locationType,
          sinNumber: site.SINNumber || site.sinNumber || site.SinNumber,
          buildingName: site.BuildingName || site.buildingName,
          numberandStreet: site.NumberandStreet || site.numberandStreet,
          villageOrSuburb: site.VillageOrSuburb || site.villageOrSuburb,
          town: site.Town || site.town,
          county: site.County || site.county,
          postcode: site.Postcode || site.postcode,
          telephoneNumber: site.TelephoneNumber || site.telephoneNumber,
          contractStartDate: site.ContractStartDate || site.contractStartDate,
          contractEndDate: site.ContractEndDate || site.contractEndDate,
          details: site.Details || site.details,
          siteSurveyComplete: site.SiteSurveyComplete || site.siteSurveyComplete,
          assignmentInstructionsIssued: site.AssignmentInstructionsIssued || site.assignmentInstructionsIssued,
          riskAssessmentIssued: site.RiskAssessmentIssued || site.riskAssessmentIssued,
          recordIsDeletedYN: site.RecordIsDeletedYN ?? site.recordIsDeletedYN ?? false,
          dateCreated: site.DateCreated || site.dateCreated || new Date().toISOString(),
          createdBy: site.CreatedBy || site.createdBy || '',
          dateModified: site.DateModified || site.dateModified,
          modifiedBy: site.ModifiedBy || site.modifiedBy
        }))
        
        console.log('✅ [SiteService] Successfully fetched sites for customer:', mappedSites.length)
        return {
          success: true,
          data: mappedSites
        }
      } else {
        console.error('❌ [SiteService] Failed to fetch sites for customer:', response.data.message)
        return {
          success: false,
          data: []
        }
      }
    } catch (error) {
      console.error('❌ [SiteService] Error fetching sites for customer:', error)
      return {
        success: false,
        data: []
      }
    }
  },

  // Get sites by region
  async getSitesByRegion(regionId: number): Promise<{ success: boolean; data: Site[] }> {
    try {
      console.log('🔄 [SiteService] Fetching sites for region:', regionId)
      
      const response = await api.get<ApiResponse<Site[]>>(SITE_ENDPOINTS.BY_REGION(regionId.toString()))
      
      if (response.data.success) {
        console.log('✅ [SiteService] Successfully fetched sites for region:', response.data.data.length)
        return {
          success: true,
          data: response.data.data
        }
      } else {
        console.error('❌ [SiteService] Failed to fetch sites for region:', response.data.message)
        return {
          success: false,
          data: []
        }
      }
    } catch (error) {
      console.error('❌ [SiteService] Error fetching sites for region:', error)
      return {
        success: false,
        data: []
      }
    }
  },

  // Get site by ID
  async getSiteById(id: number): Promise<{ success: boolean; data?: Site }> {
    try {
      console.log('🔄 [SiteService] Fetching site by ID:', id)
      
      const response = await api.get<ApiResponse<Site>>(SITE_ENDPOINTS.DETAIL(id.toString()))
      
      if (response.data.success) {
        console.log('✅ [SiteService] Successfully fetched site:', response.data.data)
        return {
          success: true,
          data: response.data.data
        }
      } else {
        console.error('❌ [SiteService] Failed to fetch site:', response.data.message)
        return {
          success: false
        }
      }
    } catch (error) {
      console.error('❌ [SiteService] Error fetching site:', error)
      return {
        success: false
      }
    }
  },

  // Create new site
  async createSite(data: CreateSiteData): Promise<{ success: boolean; data?: Site; message?: string }> {
    try {
      console.log('🔄 [SiteService] Creating site:', data)
      
      const response = await api.post<ApiResponse<Site>>(SITE_ENDPOINTS.CREATE, data)
      
      if (response.data.success) {
        console.log('✅ [SiteService] Successfully created site:', response.data.data)
        return {
          success: true,
          data: response.data.data
        }
      } else {
        console.error('❌ [SiteService] Failed to create site:', response.data.message)
        return {
          success: false,
          message: response.data.message
        }
      }
    } catch (error) {
      console.error('❌ [SiteService] Error creating site:', error)
      return {
        success: false,
        message: handleApiError(error)
      }
    }
  },

  // Update site
  async updateSite(id: number, data: UpdateSiteData): Promise<{ success: boolean; data?: Site; message?: string }> {
    try {
      console.log('🔄 [SiteService] Updating site:', id, data)
      
      const response = await api.put<ApiResponse<Site>>(SITE_ENDPOINTS.UPDATE(id.toString()), data)
      
      if (response.data.success) {
        console.log('✅ [SiteService] Successfully updated site:', response.data.data)
        return {
          success: true,
          data: response.data.data
        }
      } else {
        console.error('❌ [SiteService] Failed to update site:', response.data.message)
        return {
          success: false,
          message: response.data.message
        }
      }
    } catch (error) {
      console.error('❌ [SiteService] Error updating site:', error)
      return {
        success: false,
        message: handleApiError(error)
      }
    }
  },

  // Delete site
  async deleteSite(id: number): Promise<{ success: boolean; message?: string }> {
    try {
      console.log('🔄 [SiteService] Deleting site:', id)
      
      const response = await api.delete<ApiResponse<object>>(SITE_ENDPOINTS.DELETE(id.toString()))
      
      if (response.data.success) {
        console.log('✅ [SiteService] Successfully deleted site')
        return {
          success: true
        }
      } else {
        console.error('❌ [SiteService] Failed to delete site:', response.data.message)
        return {
          success: false,
          message: response.data.message
        }
      }
    } catch (error) {
      console.error('❌ [SiteService] Error deleting site:', error)
      return {
        success: false,
        message: handleApiError(error)
      }
    }
  }
}
